src/duct.py

```diff
@@ -164,62 +166,165 @@ def create_and_parse_args():
+
+    if capture_outputs in ["all", "stderr"] and outputs in ["all", "stderr"]:
+        stderr = TeeStream(f"{output_prefix}/stderr.txt")
+        stderr.start()
+    elif capture_outputs in ["none", "stdout"] and outputs in [
+        "all",
+        "stderr",
+    ]:
+        stderr = subprocess.PIPE
+    else:
+        stderr = subprocess.DEVNULL
+    return stdout, stderr
+
+
 def main():
     """A wrapper to execute a command, monitor and log the process details."""
     args = create_and_parse_args()
     os.makedirs(args.output_prefix, exist_ok=True)
-    try:
-        process = subprocess.Popen(
-            [str(args.command)] + args.arguments.copy(),
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            preexec_fn=os.setsid,
+    stdout, stderr = prepare_outputs(
+        args.capture_outputs, args.outputs, args.output_prefix
+    )
+    process = subprocess.Popen(
+        [str(args.command)] + args.arguments.copy(),
+        stdout=stdout,
+        stderr=stderr,
+        preexec_fn=os.setsid,
+    )
+    session_id = os.getsid(process.pid)  # Get session ID of the new process
+    report = Report(args.command, session_id)
+    report.collect_environment()
+    report.get_system_info()
+
+    if args.record_types in ["all", "processes-samples"]:
+        monitoring_args = [
+            stdout,
+            stderr,
+            report,
+            process,
+            args.report_interval,
+            args.sample_interval,
+            args.output_prefix,
+        ]
+        monitoring_thread = threading.Thread(
+            target=monitor_process, args=monitoring_args
         )
-        session_id = os.getsid(process.pid)  # Get session ID of the new process
-        report = Report(args.command, session_id)
-        report.collect_environment()
-        report.get_system_info()
-
-        while True:
-            elapsed_time = time.time() - report.start_time
-            report.collect_sample()
-            if elapsed_time >= (report.number + 1) * args.report_interval:
-                aggregated = report.aggregate_samples()
-                for pid, pinfo in aggregated.items():
-                    with open(
-                        f"{args.output_prefix}/{pid}_resource_usage.json", "a"
-                    ) as resource_statistics_log:
-                        pinfo["elapsed_time"] = elapsed_time
-                        resource_statistics_log.write(json.dumps(aggregated))
-                report.number += 1
-
-            if process.poll() is not None:  # the passthrough command has finished
-                break
-            time.sleep(args.sample_interval)
+        monitoring_thread.start()
+        monitoring_thread.join()
 
+    if args.record_types in ["all", "system-summary"]:
         with open(
-            f"{args.output_prefix}/system-report.session-{report.session_id}.json", "a"
+            f"{args.output_prefix}/system-report.session-{report.session_id}.json",
```

Comment:
following above -- might be not a directory, no need to make long filenames. best to keep session in common prefix unless you see multiple sessions for the same prefix, do you? if not do smth like

```suggestion
            f"{output_prefix_formatted}info.json",
```

Reply:
Ill resolve for now, we can open a new comment based on the changes.